File: agents/agent-server.js
const express = require('express');
const http = require('http');
const WebSocket = require('ws');

const app = express();
const server = http.createServer(app);
const wss = new WebSocket.Server({ server });

app.use(express.json());

// Message Types
const MessageType = {
    TaskRequest: 'TaskRequest',
    TaskResponse: 'TaskResponse',
    HandoffNotification: 'HandoffNotification',
    ClarificationRequest: 'ClarificationRequest',
    ClarificationResponse: 'ClarificationResponse',
    CallbackNotification: 'CallbackNotification',
    ErrorMessage: 'ErrorMessage'
};

// Server-Side Transport Abstraction
class Transport {
    constructor() {
        this.handlers = {};
    }

    on(messageType, handler) {
        this.handlers[messageType] = handler;
    }

    handleMessage(message) {
        const handler = this.handlers[message.type];
        if (handler) {
            handler(message);
        } else {
            console.error(`No handler for message type: ${message.type}`);
        }
    }

    sendResponse(client, response) {
        if (client instanceof WebSocket) {
            client.send(JSON.stringify(response));
        } else {
            client.json(response);
        }
    }
}

const transport = new Transport();

// Task Handlers
transport.on(MessageType.TaskRequest, (message) => {
    // Implement task request handling logic here
    console.log('Received TaskRequest:', message);
    // Example response:
    transport.sendResponse(message.client, {
        type: MessageType.TaskResponse,
        taskId: message.taskId,
        result: 'Task processed successfully',
        status: 'success',
        agentId: 'agent-001'
    });
});

transport.on(MessageType.ClarificationResponse, (message) => {
    // Implement clarification response handling logic here
    console.log('Received ClarificationResponse:', message);
    // Example: resume task execution
});

transport.on(MessageType.CallbackNotification, (message) => {
    // Implement callback notification handling logic here
    console.log('Received CallbackNotification:', message);
    // Example: forward result to originating agent or callback URL
});

// Server Transport Management
wss.on('connection', (ws) => {
    ws.on('message', (message) => {
        const parsedMessage = JSON.parse(message);
        parsedMessage.client = ws;
        transport.handleMessage(parsedMessage);
    });
});

// HTTP endpoints
app.post('/task', (req, res) => {
    const message = req.body;
    message.client = res;
    transport.handleMessage(message);
});

app.post('/clarification', (req, res) => {
    const message = req.body;
    message.client = res;
    transport.handleMessage(message);
});

app.post('/handoff', (req, res) => {
    const message = req.body;
    message.client = res;
    transport.handleMessage(message);
});

app.post('/callback', (req, res) => {
    const message = req.body;
    message.client = res;
    transport.handleMessage(message);
});

// Start the server
const PORT = process.env.PORT || 3000;
server.listen(PORT, () => {
    console.log(`Server running on port ${PORT}`);
});

File: agents/AgentServer/AgentServer.cs
using System;
using System.Net;
using System.Net.WebSockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace AgentServer
{
    public class AgentServer
    {
        private const int DefaultPort = 4242;
        private HttpListener _httpListener;
        private CancellationTokenSource _cancellationTokenSource;

        public virtual void Start()
        {
            _httpListener = new HttpListener();
            _httpListener.Prefixes.Add($"http://localhost:{DefaultPort}/");
            _cancellationTokenSource = new CancellationTokenSource();

            _httpListener.Start();
            Console.WriteLine($"AgentServer started on port {DefaultPort}");

            Task.Run(() => AcceptConnections(_cancellationTokenSource.Token));
        }

        private async Task AcceptConnections(CancellationToken cancellationToken)
        {
            while (!cancellationToken.IsCancellationRequested)
            {
                try
                {
                    HttpListenerContext context = await _httpListener.GetContextAsync();
                    if (context.Request.IsWebSocketRequest)
                    {
                        ProcessWebSocketRequest(context);
                    }
                    else
                    {
                        ProcessHttpRequest(context);
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error accepting connection: {ex.Message}");
                }
            }
        }

        private async void ProcessWebSocketRequest(HttpListenerContext context)
        {
            HttpListenerWebSocketContext webSocketContext = await context.AcceptWebSocketAsync(null);
            WebSocket webSocket = webSocketContext.WebSocket;

            var transport = new WebSocketTransport(webSocket);
            await HandleWebSocketConnection(transport);
        }

        private async Task HandleWebSocketConnection(TransportInterface transport)
        {
            var buffer = new byte[1024 * 4];
            while (transport is WebSocketTransport wsTransport && wsTransport.WebSocket.State == WebSocketState.Open)
            {
                var result = await wsTransport.WebSocket.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);
                if (result.MessageType == WebSocketMessageType.Text)
                {
                    string message = Encoding.UTF8.GetString(buffer, 0, result.Count);
                    HandleIncomingMessage(message, transport);
                }
            }
        }

        private void ProcessHttpRequest(HttpListenerContext context)
        {
            var transport = new HttpTransport(context.Response);
            string requestBody;
            using (var reader = new System.IO.StreamReader(context.Request.InputStream, context.Request.ContentEncoding))
            {
                requestBody = reader.ReadToEnd();
            }
            HandleIncomingMessage(requestBody, transport);
        }

        public virtual void HandleIncomingMessage(string message, TransportInterface transport)
        {
            try
            {
                JObject jsonMessage = JObject.Parse(message);
                string messageType = jsonMessage["type"]?.ToString();

                switch (messageType)
                {
                    case "TaskRequest":
                        HandleTaskRequest(JsonConvert.DeserializeObject<TaskRequest>(message), transport);
                        break;
                    case "HandoffNotification":
                        HandleHandoffNotification(JsonConvert.DeserializeObject<HandoffNotification>(message), transport);
                        break;
                    case "CallbackNotification":
                        HandleCallbackNotification(JsonConvert.DeserializeObject<CallbackNotification>(message), transport);
                        break;
                    case "ClarificationResponse":
                        HandleClarificationResponse(JsonConvert.DeserializeObject<ClarificationResponse>(message), transport);
                        break;
                    default:
                        Console.WriteLine($"Unknown message type: {messageType}");
                        break;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error handling incoming message: {ex.Message}");
            }
        }

        public virtual void SendResponse(object response, TransportInterface transport)
        {
            transport.SendResponse(response);
        }

        public virtual void HandleTaskRequest(TaskRequest taskRequest, TransportInterface transport)
        {
            // Override this method to implement task processing logic
            Console.WriteLine($"Received TaskRequest: {taskRequest.task_ID}");
        }

        public virtual void HandleHandoffNotification(HandoffNotification handoffNotification, TransportInterface transport)
        {
            // Override this method to implement handoff notification logic
            Console.WriteLine($"Received HandoffNotification: {handoffNotification.task_ID}");
        }

        public virtual void HandleCallbackNotification(CallbackNotification callbackNotification, TransportInterface transport)
        {
            // Override this method to implement callback notification logic
            Console.WriteLine($"Received CallbackNotification: {callbackNotification.task_ID}");
        }

        public virtual void HandleClarificationResponse(ClarificationResponse clarificationResponse, TransportInterface transport)
        {
            // Override this method to implement clarification response logic
            Console.WriteLine($"Received ClarificationResponse: {clarificationResponse.task_ID}");
        }
    }

    public class TaskRequest
    {
        public string task_ID { get; set; }
        public string task { get; set; }
        public string requesting_agent_ID { get; set; }
        public string callback_URL { get; set; }
    }

    public class TaskResponse
    {
        public string task_ID { get; set; }
        public string result { get; set; }
        public string status { get; set; }
        public string replying_agent_ID { get; set; }
    }

    public class HandoffNotification
    {
        public string task_ID { get; set; }
        public string current_agent_ID { get; set; }
        public string new_agent_ID { get; set; }
        public string reason_for_handoff { get; set; }
    }

    public class ClarificationRequest
    {
        public string task_ID { get; set; }
        public string question { get; set; }
        public string requesting_agent_id { get; set; }
    }

    public class ClarificationResponse
    {
        public string task_ID { get; set; }
        public string answer { get; set; }
        public string replying_agent_id { get; set; }
    }

    public class CallbackNotification
    {
        public string task_ID { get; set; }
        public string result { get; set; }
        public string status { get; set; }
        public string requesting_agent_ID { get; set; }
        public string replying_agent_ID { get; set; }
    }

    public class ErrorMessage
    {
        public string task_ID { get; set; }
        public string error_message { get; set; }
        public string requesting_agent_ID { get; set; }
    }

    public interface TransportInterface
    {
        void SendResponse(object response);
    }

    public class WebSocketTransport : TransportInterface
    {
        public WebSocket WebSocket { get; }

        public WebSocketTransport(WebSocket webSocket)
        {
            WebSocket = webSocket;
        }

        public async void SendResponse(object response)
        {
            string jsonResponse = JsonConvert.SerializeObject(response);
            byte[] buffer = Encoding.UTF8.GetBytes(jsonResponse);
            await WebSocket.SendAsync(new ArraySegment<byte>(buffer), WebSocketMessageType.Text, true, CancellationToken.None);
        }
    }

    public class HttpTransport : TransportInterface
    {
        private readonly HttpListenerResponse _response;

        public HttpTransport(HttpListenerResponse response)
        {
            _response = response;
        }

        public void SendResponse(object response)
        {
            string jsonResponse = JsonConvert.SerializeObject(response);
            byte[] buffer = Encoding.UTF8.GetBytes(jsonResponse);
            _response.ContentLength64 = buffer.Length;
            _response.ContentType = "application/json";
            _response.OutputStream.Write(buffer, 0, buffer.Length);
            _response.Close();
        }
    }
}